File: src/storage/ref_store.cpp
#include "flux/storage/ref_store.hpp"
#include "flux/util/filesystem.hpp"
#include <fmt/format.h>
#include <fstream>

namespace flux {

RefStore::RefStore(Path refs_dir) : refs_dir_(std::move(refs_dir)) {
}

Result<ObjectId> RefStore::read(const std::string& name) {
    Path ref_path = get_ref_path(name);
    
    auto data_result = Filesystem::read_file(ref_path);
    if (!data_result) {
        return flux::unexpected(data_result.error());
    }
    
    // Parse object ID from file content
    std::string_view content(reinterpret_cast<const char*>(data_result->data()), data_result->size());
    
    // Trim whitespace
    while (!content.empty() && std::isspace(content.back())) {
        content.remove_suffix(1);
    }
    
    return ObjectId::from_hex(content);
}

Result<void> RefStore::write(const std::string& name, const ObjectId& id) {
    Path ref_path = get_ref_path(name);
    
    // Create parent directories
    auto dir_result = Filesystem::create_directories(ref_path.parent_path());
    if (!dir_result) {
        return dir_result;
    }
    
    // Write object ID
    std::string content = id.to_hex() + "\n";
    Bytes data(content.begin(), content.end());
    
    return Filesystem::write_file(ref_path, data);
}

Result<void> RefStore::remove(const std::string& name) {
    return Filesystem::remove(get_ref_path(name));
}

bool RefStore::exists(const std::string& name) {
    return Filesystem::exists(get_ref_path(name));
}

Result<std::vector<std::string>> RefStore::list() {
    std::vector<std::string> refs;
    
    // List refs/heads/
    Path heads_dir = refs_dir_ / "heads";
    if (Filesystem::exists(heads_dir)) {
        auto heads_result = Filesystem::list_directory(heads_dir);
        if (!heads_result) {
            return flux::unexpected(heads_result.error());
        }
        
        for (const auto& path : *heads_result) {
            refs.push_back("refs/heads/" + path.filename().string());
        }
    }
    
    // List refs/tags/
    Path tags_dir = refs_dir_ / "tags";
    if (Filesystem::exists(tags_dir)) {
        auto tags_result = Filesystem::list_directory(tags_dir);
        if (!tags_result) {
            return flux::unexpected(tags_result.error());
        }
        
        for (const auto& path : *tags_result) {
            refs.push_back("refs/tags/" + path.filename().string());
        }
    }
    
    return refs;
}

Result<std::string> RefStore::read_symbolic(const std::string& name) {
    Path ref_path = refs_dir_.parent_path() / name;  // e.g., .flux/HEAD
    
    auto data_result = Filesystem::read_file(ref_path);
    if (!data_result) {
        return flux::unexpected(data_result.error());
    }
    
    std::string_view content(reinterpret_cast<const char*>(data_result->data()), data_result->size());
    
    // Format: "ref: refs/heads/main\n"
    if (!content.starts_with("ref: ")) {
        return flux::unexpected("Not a symbolic reference");
    }
    
    content.remove_prefix(5);  // Remove "ref: "
    
    // Trim whitespace
    while (!content.empty() && std::isspace(content.back())) {
        content.remove_suffix(1);
    }
    
    return std::string(content);
}

Result<void> RefStore::write_symbolic(const std::string& name, const std::string& target) {
    Path ref_path = refs_dir_.parent_path() / name;
    
    std::string content = fmt::format("ref: {}\n", target);
    Bytes data(content.begin(), content.end());
    
    return Filesystem::write_file(ref_path, data);
}

Path RefStore::get_ref_path(const std::string& name) const {
    // name is like "refs/heads/main"
    return refs_dir_.parent_path() / name;
}

} // namespace flux

File: src/storage/ref_store.zig
const std = @import("std");
const types = @import("../core/types.zig");
const object_id_module = @import("../core/object_id.zig");
const filesystem = @import("../util/filesystem.zig");

const ObjectId = object_id_module.ObjectId;
const Filesystem = filesystem.Filesystem;

/// RefStore - Manages references (branches, tags, HEAD)
pub const RefStore = struct {
    refs_dir: []const u8,
    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator, refs_dir: []const u8) RefStore {
        return .{
            .allocator = allocator,
            .refs_dir = refs_dir,
        };
    }

    /// Read reference
    pub fn read(self: *RefStore, name: []const u8) !ObjectId {
        const ref_path = try self.getRefPath(name);
        defer self.allocator.free(ref_path);

        const data = try Filesystem.readFile(self.allocator, ref_path);
        defer self.allocator.free(data);

        // Trim whitespace
        const content = std.mem.trim(u8, data, &std.ascii.whitespace);

        return try ObjectId.fromHex(self.allocator, content);
    }

    /// Write reference
    pub fn write(self: *RefStore, name: []const u8, id: *const ObjectId) !void {
        const ref_path = try self.getRefPath(name);
        defer self.allocator.free(ref_path);

        const content = try id.toHex(self.allocator);
        defer self.allocator.free(content);

        var data = std.ArrayList(u8).init(self.allocator);
        defer data.deinit();

        try data.appendSlice(content);
        try data.append('\n');

        try Filesystem.writeFile(ref_path, data.items);
    }

    /// Remove reference
    pub fn remove(self: *RefStore, name: []const u8) !void {
        const ref_path = try self.getRefPath(name);
        defer self.allocator.free(ref_path);
        try Filesystem.remove(ref_path);
    }

    /// Check if reference exists
    pub fn exists(self: *RefStore, name: []const u8) !bool {
        const ref_path = try self.getRefPath(name);
        defer self.allocator.free(ref_path);
        return Filesystem.exists(ref_path);
    }

    /// List all references
    pub fn list(self: *RefStore) ![][]const u8 {
        var refs = std.ArrayList([]const u8).init(self.allocator);
        errdefer {
            for (refs.items) |ref| self.allocator.free(ref);
            refs.deinit();
        }

        // List refs/heads/
        const heads_dir = try std.fs.path.join(self.allocator, &[_][]const u8{ self.refs_dir, "heads" });
        defer self.allocator.free(heads_dir);

        if (Filesystem.exists(heads_dir)) {
            const entries = try Filesystem.listDir(self.allocator, heads_dir);
            defer {
                for (entries) |entry| self.allocator.free(entry);
                self.allocator.free(entries);
            }

            for (entries) |entry| {
                const ref_name = try std.fmt.allocPrint(self.allocator, "refs/heads/{s}", .{entry});
                try refs.append(ref_name);
            }
        }

        // List refs/tags/
        const tags_dir = try std.fs.path.join(self.allocator, &[_][]const u8{ self.refs_dir, "tags" });
        defer self.allocator.free(tags_dir);

        if (Filesystem.exists(tags_dir)) {
            const entries = try Filesystem.listDir(self.allocator, tags_dir);
            defer {
                for (entries) |entry| self.allocator.free(entry);
                self.allocator.free(entries);
            }

            for (entries) |entry| {
                const ref_name = try std.fmt.allocPrint(self.allocator, "refs/tags/{s}", .{entry});
                try refs.append(ref_name);
            }
        }

        return refs.toOwnedSlice();
    }

    /// Read symbolic reference (e.g., HEAD)
    pub fn readSymbolic(self: *RefStore, name: []const u8) ![]u8 {
        const parts = [_][]const u8{ self.refs_dir, "..", name };
        const ref_path = try std.fs.path.join(self.allocator, &parts);
        defer self.allocator.free(ref_path);

        const data = try Filesystem.readFile(self.allocator, ref_path);
        defer self.allocator.free(data);

        var content = std.mem.trim(u8, data, &std.ascii.whitespace);

        // Format: "ref: refs/heads/main"
        if (!std.mem.startsWith(u8, content, "ref: ")) {
            return error.InvalidReference;
        }

        return try self.allocator.dupe(u8, content[5..]);
    }

    /// Write symbolic reference
    pub fn writeSymbolic(self: *RefStore, name: []const u8, target: []const u8) !void {
        const parts = [_][]const u8{ self.refs_dir, "..", name };
        const ref_path = try std.fs.path.join(self.allocator, &parts);
        defer self.allocator.free(ref_path);

        const content = try std.fmt.allocPrint(self.allocator, "ref: {s}\n", .{target});
        defer self.allocator.free(content);

        try Filesystem.writeFile(ref_path, content);
    }

    fn getRefPath(self: *RefStore, name: []const u8) ![]u8 {
        const parts = [_][]const u8{ self.refs_dir, "..", name };
        return try std.fs.path.join(self.allocator, &parts);
    }
};
